Add check for skipping in When Expressions with Result variable replacement

jerop opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

<!-- Describe your changes here- ideally you can get that description straight from
your descriptive commit message(s)! -->

tasks with when expressions using variables that evaluated to false were mistakenly executed because it didn't have an additional check to validate that the task should be skipped after task result variable replacement

the point where we excluded the tasks with when expressions containing static inputs or parameter variables (which are replaced earlier as read from the pipeline) is here: https://github.com/tektoncd/pipeline/blob/079a6c8bce6eb5663689fad7f50d0fce7591e6bc/pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go#L260-L272


at that point, when expressions containing results don't have the variables replaced so they are not excluded -- that function is used here: 
https://github.com/tektoncd/pipeline/blob/906414229d999cd9c7e6a5401f937c0bb3d97e7f/pkg/reconciler/pipelinerun/pipelinerun.go#L499-L511

the task results are applied later, right after which we add this additional check to exclude the when expressions containing results: https://github.com/tektoncd/pipeline/blob/906414229d999cd9c7e6a5401f937c0bb3d97e7f/pkg/reconciler/pipelinerun/pipelinerun.go#L522-L527

Fixes https://github.com/tektoncd/pipeline/issues/3188

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [x] Release notes block has been filled in or deleted (only if no user facing changes)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

# Release Notes

```release-note
`Tasks` with `WhenExpressions` using variable replacements are not executed when the `WhenExpressions` evaluate to false
```

<!--
Describe any user facing changes here, or delete this block.

Examples of user facing changes:
- API changes
- Bug fixes
- Any changes in behavior
- Changes requiring upgrade notices or deprecation warnings

For pull requests with a release note:

    ```release-note
    Your release note here
    ```

For pull requests that require additional action from users switching to the new release, include the string "action required" (case insensitive) in the release note:

    ```release-note
    action required: your release note here
    ```

For pull requests that don't need to be mentioned at release time, use the `/release-note-none` Prow command to add the `release-note-none` label to the PR. You can also write the string "NONE" as a release note in your PR description:

    ```release-note
    NONE
    ```
-->

bobcatfish commented on examples/v1beta1/pipelineruns/pipelinerun-with-when-expressions.yaml:
```diff
@@ -66,6 +66,16 @@ spec:
             script: 'echo file exists'
+    - name: echo-file-missing
+      when:
+        - input: "$(tasks.check-file.results.status)"
```
i wonder if we should call this something other than "status" to reduce confusion around being able to execute if a previous task fails

jerop replied:
made the update on the naming in this open pr: https://github.com/tektoncd/pipeline/pull/3194

should i move the naming change here?

bobcatfish commented:
Can you confirm that the example you updated fails without your change?

I also want to check, #3196 implies this is finally related but it sounds like that's a red herring and it's actually related to variable replacement?

(I'm wondering also if makes sense to have an example/test that verifies that finally works properly with this as well)

jerop commented:
/test tekton-pipeline-unit-tests

pritidesai commented:
@jerop, please add a unit test in `pipelinerun_test.go`. I just modified an existing unit test to not match the task result value (resulting when expression to `false`) and that task was skipped.

I am trying to reproduce it through the integration test from examples.

jerop commented:
@pritidesai thank you for the review! 

added a task to the when expressions example that reproduced the issue, and also added checks in the unit tests to ensure the taskruns are not created when variable replacement is used -- please have a look when you have a chance

pritidesai commented:
/kind bug

pritidesai commented on pkg/reconciler/pipelinerun/pipelinerun.go:
```diff
@@ -522,7 +522,7 @@ func (c *Reconciler) runNextSchedulableTask(ctx context.Context, pr *v1beta1.Pip
 
 	for _, rprt := range nextRprts {
-		if rprt == nil {
+		if rprt == nil || rprt.Skip(pipelineState, d) {
```
This looks fine for now but would like to see this being evaluated as an attribute to `WhenExpressions` similar to what we have for the parameter. That way, all the checks are in one place irrespective of param mismatch or task result mismatch, a task and all its dependencies are guaranteed to be skipped.

jerop replied:
the point where we excluded the tasks with when expressions containing static inputs or parameter variables (which are replaced earlier as read from the pipeline) is here: https://github.com/tektoncd/pipeline/blob/079a6c8bce6eb5663689fad7f50d0fce7591e6bc/pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go#L260-L272


at that point, when expressions containing results don't have the variables replaced so they are not excluded -- that function is used here: 
https://github.com/tektoncd/pipeline/blob/906414229d999cd9c7e6a5401f937c0bb3d97e7f/pkg/reconciler/pipelinerun/pipelinerun.go#L499-L511


the task results are applied later right before this additional check to exclude the when expressions containing results: https://github.com/tektoncd/pipeline/blob/906414229d999cd9c7e6a5401f937c0bb3d97e7f/pkg/reconciler/pipelinerun/pipelinerun.go#L522-L527

the options I see are:

1. removing the earlier check, so that we have only one check after task results have been applied -- but this means will be processing more tasks than necessary which could have been excluded earlier
2. as we have it now, do two checks with the first one reducing the list next tasks and reducing unnecessary work -- I'm leaning towards this but can see the benefit of doing the check once

please let me know what you think

pritidesai replied:
@jerop I am with you - option 2 is fine.

But I was thinking more towards, adding call to `ApplyTaskResults` in `when_types.go` and update `AllowsExecution` to deal with constants, params, and task results. `AllowsExecution` in the end would be responsible for checking if constants match, resolving params/task results and check if they match. Similar to what we do with `applyReplacements` and `GetVarSubsutitionExpressions` as attributes to `WhenExpression`.

These changes does not have to be part of this PR. It's more about refactoring which can be done in a separate PR.

jerop replied:
thanks for the clarification @pritidesai -- will do that refactor soon

pritidesai commented on pkg/reconciler/pipelinerun/pipelinerun_test.go:
```diff
@@ -2081,6 +2081,21 @@ func TestReconcileWithWhenExpressionsWithParameters(t *testing.T) {
+	}
 }
 
 func TestReconcileWithWhenExpressionsWithTaskResults(t *testing.T) {
```
@jerop I am expecting a pipeline task here which has reference to task result but the result does not match with what is expected i.e.

```
		tb.PipelineTask("task-result-mismatch", "task",
			tb.PipelineTaskWhenExpression("$(tasks.a-task.results.aResult)", selection.In, []string{"random"}),
		),
```

pritidesai replied:
And also please add a test, checking the pipeline skips the entire branch in case of a task result mismatch.

jerop replied:
@pritidesai the example with result reference doesn't match what is expected is already here (c-task): https://github.com/tektoncd/pipeline/pull/3197/files/906414229d999cd9c7e6a5401f937c0bb3d97e7f#diff-f7c8532cfe79319c0fff711a4979a6d6R2107

and it has a dependent (d-task) that's also skipped https://github.com/tektoncd/pipeline/pull/3197/files/906414229d999cd9c7e6a5401f937c0bb3d97e7f#diff-f7c8532cfe79319c0fff711a4979a6d6R2112

bobcatfish commented on examples/v1beta1/pipelineruns/pipelinerun-with-when-expressions.yaml:
```diff
@@ -66,6 +66,16 @@ spec:
           - name: echo
             image: ubuntu
             script: 'echo file exists'
+    - name: echo-file-missing
```
i think it would be worth adding a couple comments or changing the names of the pipelinetasks to reflect what we are trying to cover here, which i think is:

1. using a task result which causes the when expression to fail (echo-file-missing)
2. using a pipeline param which causes the when expression to fail (task-should-be-skipped)

This example doesnt seem to include any cases where the when expression succeeds 🤔  cant remember if there was another exmaple? i think we should add that as well if not

but we dont have to do that in this pr!

jerop replied:
sounds good, will update the names/add comments 

`echo-file-exists` is an example using when expressions with results and succeeds -- it uses the same result `$(tasks.check-file.results.exists)` and evaluates to true:
https://github.com/tektoncd/pipeline/blob/18dd0af6abf304c18b1484c52cf3b75c7e985160/examples/v1beta1/pipelineruns/pipelinerun-with-when-expressions.yaml#L59-L78

`create-file` is the example that uses when expressions with params and succeeds:
https://github.com/tektoncd/pipeline/blob/18dd0af6abf304c18b1484c52cf3b75c7e985160/examples/v1beta1/pipelineruns/pipelinerun-with-when-expressions.yaml#L16-L31

bobcatfish replied:
whoops, sorry i missed that @jerop !

bobcatfish reviewed (commented):
I agree with @pritidesai about refactoring the code a bit - and @jerop with your analysis of trying to find a way to only have to check once so we dont have to worry about checking for skipped tasks correctly in two different places!

but let's tackle that after we get this fixed :D

/lgtm
/approve
